From b8fee0a8ff5c2eae224ecabff4d40303bde3c1f3 Mon Sep 17 00:00:00 2001
From: Arpit Agarwal <arp@apache.org>
Date: Mon, 12 Feb 2018 07:09:54 -0800
Subject: [PATCH 2776/2863] HDFS-10453. ReplicationMonitor thread could stuck
 for long time due to the race between replication
 and delete of same file in a large cluster..
 Contributed by He Xiaoqiao.

(cherry picked from commit 02f6030b35999f2f741a8c4b9363ee59f36f7e28)

 Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java

Change-Id: Ie13129e0f69eb4f61e0db4c20e05bb670e59a92f
---
 .../hdfs/server/blockmanagement/BlockManager.java  |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java
index 8aef73a..9c05e50 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java
@@ -4053,6 +4053,7 @@ public void clearQueues() {
   private static class ReplicationWork {
     private final BlockInfo block;
     private final String srcPath;
+    private final long blockSize;
     private final byte storagePolicyID;
     private final DatanodeDescriptor srcNode;
     private final int additionalReplRequired;
@@ -4070,6 +4071,7 @@ public ReplicationWork(BlockInfo block,
         int priority) {
       this.block = block;
       this.srcPath = bc.getName();
+      this.blockSize = block.getNumBytes();
       this.storagePolicyID = bc.getStoragePolicyID();
       this.srcNode = srcNode;
       this.srcNode.incrementPendingReplicationWithoutTargets();
@@ -4086,7 +4088,7 @@ private void chooseTargets(BlockPlacementPolicy blockplacement,
       try {
         targets = blockplacement.chooseTarget(getSrcPath(),
             additionalReplRequired, srcNode, liveReplicaStorages, false,
-            excludedNodes, block.getNumBytes(),
+            excludedNodes, blockSize,
             storagePolicySuite.getPolicy(getStoragePolicyID()), null);
       } finally {
         srcNode.decrementPendingReplicationWithoutTargets();
-- 
1.7.9.5

